Add bme68x lib

nkanderson opened the pull request:
Bringing the Bosch BME68x sensor driver over from the [old repo](https://github.com/coffee-and-telesense/PicoAPRS-Firmware), attempting to make the fewest changes possible at this stage in order to produce a successful build. Relevant output from the cmake build in VSCode is showing below:

```sh
[build] [23/23] Linking C executable tests/bme68x_test/bme68x_test.elf
[build] Memory region         Used Size  Region Size  %age Used
[build]              RAM:        2736 B        32 KB      8.35%
[build]            FLASH:       40636 B       256 KB     15.50%
[driver] Build completed: 00:00:00.952
[build] Build finished with exit code 0
```

The main changes from the original include the following:

**Driver changes**
- Changing the `include` for the stm32 HAL in `bme68x_driver.h` from `stm32l4xx_hal_i2c.h` to `stm32u0xx_hal.h`, which obviously changes the family from L4 to U0, but also references the entire HAL instead of just the I2C component. The build was failing when using just the I2C component for U0, probably due to changes in our overall build process. I think this is still correct though (maybe more correct than previously).
- Need to include `string.h` in `bme68x_driver.h` for `memset`. Again, probably due to how we were building in the previous repo, but it seems like if this is needed for `memset`, then including it explicitly should be ok. We can certainly revisit this though if needed.

**Test application changes**
- We don't currently have the same logging utility, so I added the basic `debug_print` directly to the test file for now. We should decide if we want to bring the logging utility over as-is, or move to a different logging implementation.
- I just commented out the blinky bit for now. I'll likely add that back as it is sort of a nice confirmation that the program is running when troubleshooting, but it seems fine to leave it as-is for the time being.

**Delay function changes**
- I was running into a build error where `HAL_TIM_PeriodElapsedCallback` had duplicate definitions. I'm not sure if the HAL changed in this regard, but based on examination of the existing `HAL_TIM_PeriodElapsedCallback`, it seemed that the appropriate thing to do was to implement a `app_timer_elapsed_hook`, which would be called from the HAL callback.

nkanderson committed:
Add Bosch BME68x sensor driver

This commit copies the BME68x driver from the original PicoAPRS-Firmware
repo to this one, with the minimal changes necessary to produce a
successful build.

nkanderson committed:
Add bme68x library and test app to base cmake config

nkanderson committed:
Add basic BME68x test application
This test application will ideally be cleaned up to update LED references and consolidate logging functionality as desired. That work may be done after validating this basic test application is working.

nkanderson committed:
Add delay function to timer code for working with BME68x driver

nkanderson commented:
Leaving this as a draft for now, because I'm in an unfortunate position of being unable to track down a USBC cable with data transfer. So I've confirmed the build succeeds, but I haven't been able to run it on the nucleo U0 board yet with the sensor. I did want to open this request up though so we could talk through some things like the logging utility and our approach for things like the delay function.

Main points of discussion being:
- Do we want to bring the logging utility over from the old repo, or use something else?
- Is there a different approach we want to take for something like the required delay function? Or is this fine for the test application?

@reecewayt let me know if you want to meet up and talk through any of this in real time.

reecewayt commented:
### Timer Question: 
Let's definitely sit down and chat sometime, but overall, this looks good to me. You are correct for using the app timer hook. 

Basically, the system clock is currently being used by threadX, because the RTOS needs a timer for scheduling and counting ticks, etc. Since threadX takes over the system clock, I had to configure the system clock to use TIM6 in CubeMX. In this sense, RTOS time and System Time are different. 

![image](https://github.com/user-attachments/assets/8a32cb13-181e-4b1c-ae2f-1591fa1e7ced)
 
This is why you see that CubeMX generated the following code, and I added the app_timer_elapsed_hook() because I had the same issue as you did with the multiple definitions error:
```
/**
  * @brief  Period elapsed callback in non blocking mode
  * @note   This function is called  when TIM6 interrupt took place, inside
  * HAL_TIM_IRQHandler(). It makes a direct call to HAL_IncTick() to increment
  * a global variable "uwTick" used as application time base.
  * @param  htim : TIM handle
  * @retval None
  */
void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
{
  /* USER CODE BEGIN Callback 0 */

  /* USER CODE END Callback 0 */
  if (htim->Instance == TIM6) {
    HAL_IncTick();
  }
  /* USER CODE BEGIN Callback 1 */
  // Call the application hook - this is where custom code is injected
  else {
    app_timer_elapsed_hook(htim);
  }
  /* USER CODE END Callback 1 */
}
```

Anyway, it's a good find on your part, and I'm sorry I didn't have documentation out about this.

### Logging question: 

I have implemented a slightly different logging utility that utilizes a printf redirection. Here's the header file, and I meant to commit it but haven't gotten around to it. I just learned a cool git trick so I'll do that now :) . 

```c
/*******************************************************************************
 * @file: logging.h
 * @brief: Simple debug printing utility for STM32 projects using printf redirection
 *
 * @note: Requires UART to be initialized before use
 ******************************************************************************/
#pragma once

#include "stm32u0xx_hal.h"
#include "usart.h"
#include <stdio.h>

// Define printf redirection prototype
#define PUTCHAR_PROTOTYPE int __io_putchar(int ch)

// UART handle to be registered by the application
extern UART_HandleTypeDef *debug_uart_handle;

// Function to initialize the debug logger with a specific UART handle
void debug_init(UART_HandleTypeDef *huart);

// Debug logging macros
#ifdef DEBUG
    #include <string.h>
    #include <stdarg.h>
    
    // Main debug print macro that will be active in DEBUG builds
    #define DEBUG_PRINT(fmt, ...) printf(fmt, ##__VA_ARGS__)
    
    // Additional debug macros with log levels if needed
    #define DEBUG_INFO(fmt, ...)  printf("[INFO] " fmt, ##__VA_ARGS__)
    #define DEBUG_WARN(fmt, ...)  printf("[WARN] " fmt, ##__VA_ARGS__)
    #define DEBUG_ERROR(fmt, ...) printf("[ERROR] " fmt, ##__VA_ARGS__)

    // Conditional debug printing (only prints if condition is true)
    #define DEBUG_IF(cond, fmt, ...) do { if (cond) printf(fmt, ##__VA_ARGS__); } while(0)
    
    // Debug printing with function name and line number information
    #define DEBUG_TRACE(fmt, ...) printf("[%s:%d] " fmt, __FUNCTION__, __LINE__, ##__VA_ARGS__)
    
#else
    // Empty macros when DEBUG is not defined
    #define DEBUG_PRINT(fmt, ...) ((void)0)
    #define DEBUG_INFO(fmt, ...)  ((void)0)
    #define DEBUG_WARN(fmt, ...)  ((void)0)
    #define DEBUG_ERROR(fmt, ...) ((void)0)
    #define debug_print(fmt, ...) ((void)0)
#endif

// Debug logger initialization function is always defined
// but only does something when DEBUG is defined
void debug_init(UART_HandleTypeDef *huart);

// *********************Example usage:************************************************
/**
void initialize_sensors(void) {
    DEBUG_TRACE("Starting sensor initialization\r\n");
    // Function body...
}
// Output: 
// [initialize_sensors:3] Starting sensor initialization

void read_sensor_data(void) {
    DEBUG_INFO("Reading sensor data\r\n");
    // Function body...
}
// Output:
// [INFO] Reading sensor data


DEBUG_IF(temperature > 30, "Temperature alert: %d°C\r\n", temperature);
// Output: (Only if temperature > 30)
// Temperature alert: 35°C 


**********************************************************************************************/
```

nkanderson commented:
> Anyway, it's a good find on your part, and I'm sorry I didn't have documentation out about this.

Oh no problem, thanks for the detailed explanation. I was able to figure out the hook quickly, so I think that's a sign it was a good solution on your end.

> I have implemented a slightly different logging utility that utilizes a printf redirection.

Gotcha, this makes more sense! I saw the `printf` invocations in your test app and thought maybe there was just something I wasn't understanding about the execution context. This looks good to me, and I approved the PR. I'll plan to rebase this branch once that's merged, and make the necessary changes to my test app.

If this is otherwise looking good, I'll plan to test on my U0 nucleo board once I have a cable with data transfer (I still can't believe I don't have one lying around 🫠 ), then mark this as "ready for review" once it's confirmed to be working.

nkanderson committed:
Update BME68x test to use logging utility printf

nkanderson commented:
Ok, finally got a cable and was able to confirm this is working. I was able to switch my test code over to using the logging utility as well, so this should be ready for review @reecewayt

nkanderson commented:
@reecewayt let me know if you want a quick walk-through of this. The main difference between this and what you reviewed earlier is that I added some cmake stuff for the logging utility, and updated my test application to use that utility.
